### PR TITLE
Implementing GPS as a mechanism to incorporate global attention for all MPNNs

### PR DESCRIPTION
**Pull request by ArCho48**

### CODE REVIEW

File: hydragnn/models/Base.py [hunk 1]
```diff
@@ -119,6 +130,35 @@ def __init__(
...
+        # Specify learnable embeddings
+        if self.use_global_attn or self.use_encodings:
+            self.pos_emb = Linear(self.pe_dim, self.hidden_dim, bias=False) 
+            if self.input_dim:
+                self.node_emb = Linear(self.input_dim, self.hidden_dim, bias=False)
+                self.node_lin = Linear(2*self.hidden_dim, self.hidden_dim, bias=False)
+            if self.is_edge_model:
+                self.rel_pos_emb = Linear(self.pe_dim, self.hidden_dim, bias=False) 
+                if self.use_edge_attr:
+                    self.edge_emb = Linear(self.edge_dim, self.hidden_dim, bias=False)
```
Comment (RylieWeaver) [inline hydragnn/models/Base.py:159]: @ArCho48 

This dimensionality may cause problems for some of the models that use edge_dim for creating some of their blocks within the message-passing stack. For example, DimeNet has this line in `HydraEmbeddingBlock`:

`self.edge_lin = Linear(edge_dim, hidden_channels)`

Comment (ArCho48) [reply hydragnn/models/Base.py:159]: @RylieWeaver updated edge_dim is passed to get_conv in most stacks to create their individual GNN blocks. For ex., in dimenet I do this:
Screenshot 2024-12-02 at 10 41 27 PM

File: examples/qm9/qm9.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
 import os
+os.environ['TORCH_DISTRIBUTED_DEBUG'] = 'DETAIL'
```
Comment (allaffa) [inline examples/qm9/qm9.py:2]: @ArCho48 
Is this really needed?

File: examples/zinc/zinc.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+import pdb
+import torch
+import torch_geometric
+from torch_geometric.datasets import ZINC
+import torch_geometric.transforms as T
+from torch_geometric.transforms import AddLaplacianEigenvectorPE
+
+# torch.cuda.set_device(0)
+# os.environ['CUDA_VISIBLE_DEVICES'] = ''
+os.environ['KINETO_LOG_LEVEL'] = '3'
```
Comment (allaffa) [inline examples/zinc/zinc.py:12]: @ArCho48 
Why did you have to do this?

File: examples/zinc/zinc.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    test_loader,
+    writer,
+    scheduler,
+    config["NeuralNetwork"],
+    log_name,
+    verbosity,
+)
+
+
+# //"message_passing": "PNA",
```
Comment (allaffa) [inline examples/zinc/zinc.py:99]: Clean out this commented sections, which is clearly not needed.

File: hydragnn/globalAtt/gps.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    def forward(
+        self,
+        inv_node_feat: Tensor,
+        equiv_node_feat: Tensor,
+        batch: Optional[torch.Tensor] = None,
+        **kwargs
+    ) -> Tensor:
+        """Runs the forward pass of the module."""
+        hs = []
+        # kwargs = {"edge_index":edge_index}
```
Comment (allaffa) [inline hydragnn/globalAtt/gps.py:110]: If not needed anymore, clean out these commented lines.

File: hydragnn/globalAtt/gps.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        # if edge_attr is not None:
+        #     kwargs.update({"edge_attr":edge_attr})
+        # pdb.set_trace()
+        if self.conv is not None:  # Local MPNN.
+            h, equiv_node_feat = self.conv(inv_node_feat=inv_node_feat, equiv_node_feat=equiv_node_feat, **kwargs) 
+            # h = self.mpnn_out(h) # map back mpnn (ex., GAT) output to hidden_dim
```
Comment (allaffa) [inline hydragnn/globalAtt/gps.py:116]: If not needed, remove this commented line.

File: hydragnn/models/CGCNNStack.py [hunk 7]
```diff
@@ -23,27 +23,28 @@ def __init__(
...
         output_dim,
         *args,
         **kwargs,
     ):
         self.edge_dim = edge_dim
-
+        self.is_edge_model = True #specify that mpnn can handle edge features
         # CGCNN does not change embedding dimensions
-        # We use input dimension (first argument of base constructor)
+        # We use input dimension (first argument of base constructor)00
```
Comment (allaffa) [inline hydragnn/models/CGCNNStack.py:34]: @ArCho48 
does the "00" at the end meaning anything? if not. remove it.

File: examples/qm9/qm9.py (hunk 2 above)
Comment (ArCho48) [reply examples/qm9/qm9.py:2]: This was to debug the unused parameter error. Once all the tests check out, I will take a final pass removing all these lines

File: examples/zinc/zinc.py (hunk 3 above)
Comment (ArCho48) [reply examples/zinc/zinc.py:12]: Added the os variables at different times for debugging. Will remove them after all the tests pass

File: examples/zinc/zinc.py (hunk 4 above)
Comment (ArCho48) [reply examples/zinc/zinc.py:99]: Will do, thanks. Please note that this is not the final formatting, I will take a last pass after all the tests check out and before I run black

File: hydragnn/globalAtt/gps.py (hunk 5 above)
Comment (ArCho48) [reply hydragnn/globalAtt/gps.py:110]: Sure

File: hydragnn/globalAtt/gps.py (hunk 6 above)
Comment (ArCho48) [reply hydragnn/globalAtt/gps.py:116]: Will do, thanks

**Review by allaffa — COMMENTED** [review]
Code-wise, everything seems to be ok expect for some commented lines that are better to remove for clean-ups. 
Let us meet to see why the CI tests fails, and let's fix them.

**Pull request closed without merging**